### PR TITLE
Experiments

### PR DESCRIPTION
**Pull request by Paulina100**

### CODE REVIEW

**Review by tunczyk101 — APPROVED** [review]
Wygląda w miarę ok
Tylko te typy

File: training/results.py [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+from matplotlib import pyplot as plt
+from sklearn.metrics import (
+    ConfusionMatrixDisplay,
+    confusion_matrix,
+    mean_absolute_error,
+    mean_squared_error,
+)
+
+
+def print_check_predictions(y, y_pred):
```
[the remark this answers is not in the document]
Comment (j-adamczyk) [reply training/results.py:11]: +1

File: training/results.py [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+    print(f"MSE: {mean_squared_error(y, y_pred):.2f}")
+    print(f"RMSE: {mean_squared_error(y, y_pred, squared=False):.2f}\n")
+
+
+def round_predictions(threshold: str | float, predict):
+    """
+    Round predicted values based on a specified threshold.
+
+    If threshold is a string "round", it rounds normally.
+    If 'threshold' is a float, values above the threshold are rounded up, and values below are rounded down.
```
Comment (j-adamczyk) [inline training/results.py:27]: Czemu po prostu nie zrobić domyślnej wartości 0.5?

File: training/results.py [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+
+    # predict for train and test
+    y_pred_test = model.predict(X_test)
+    y_pred_train = model.predict(X_train)
+
+    # calculate train and test RMSE, MSE and MAE
+    rmse_train = mean_squared_error(y_train, y_pred_train, squared=False)
+    rmse_test = mean_squared_error(y_test, y_pred_test, squared=False)
+    mse_test = mean_squared_error(y_test, y_pred_test)
+    mae_test = mean_absolute_error(y_test, y_pred_test)
```
Comment (j-adamczyk) [inline training/results.py:113]: Odpuściłbym MSE, tylko RMSE i MAE

**Review by j-adamczyk — CHANGES_REQUESTED** [review]
Parę uwag (w notebookach irytująco się komentuje):
1. Wykresy muszą mieć tytuły
2. Skrót od "number" to typowo "no" albo "#"
3. Tutaj to są bardziej "features" niż "characteristics", bo jednak niektóre są dość kreatywnie zdefiniowane, np. obrażenia
4. Co to znaczy "normal" w cechach? Jeżeli to jest brak tuningu, to oznaczyłbym jako "default" albo coś w tym stylu, bo "normal" kojarzy się od razu z normalizacją cech
5. Co robiło "minimise_overfitting"?
6. Jak zapisuejsz JSONy, to polecam opcję `indent=2`, wtedy ładnie to sformatuje
7. Zmieniłbym 8, 13, 50 na jakieś zestawy cech, które opiszecie w tekście. Typu np. "basic", "expanded", "full". Bo sama liczba cech to raczej artefakt obliczeń, a human-readable nazwy bardziej oddają to, co one robią

Ogólnie - bardzo fajna analiza, bardzo solidne wykresy.

**Comment by Paulina100** [comment]
> Parę uwag (w notebookach irytująco się komentuje):
> 
> 1. Wykresy muszą mieć tytuły
> 2. Skrót od "number" to typowo "no" albo "#"
> 3. Tutaj to są bardziej "features" niż "characteristics", bo jednak niektóre są dość kreatywnie zdefiniowane, np. obrażenia
> 4. Co to znaczy "normal" w cechach? Jeżeli to jest brak tuningu, to oznaczyłbym jako "default" albo coś w tym stylu, bo "normal" kojarzy się od razu z normalizacją cech
> 5. Co robiło "minimise_overfitting"?
> 6. Jak zapisuejsz JSONy, to polecam opcję `indent=2`, wtedy ładnie to sformatuje
> 7. Zmieniłbym 8, 13, 50 na jakieś zestawy cech, które opiszecie w tekście. Typu np. "basic", "expanded", "full". Bo sama liczba cech to raczej artefakt obliczeń, a human-readable nazwy bardziej oddają to, co one robią
> 
> Ogólnie - bardzo fajna analiza, bardzo solidne wykresy.

ad. 5. Zmniejsza głębokość drzewa (max_depth) żeby się nie przeuczało. Niewiele to dało, ale myślałam że tak się robi.

File: notebooks/models/lightgbm/lightgbm_basic_results.json [hunk 4]
```diff
@@ -0,0 +1,32 @@
+{
```
Comment (j-adamczyk) [inline notebooks/models/lightgbm/lightgbm_basic_results.json:1]: W sumie czemu JSON, a nie CSV?

File: training/results.py [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+    """
+    Calculate and print RMSE for predicted values.
+
+    :param y: True values.
+    :param y_pred: Predicted values from a model.
+    """
+    print(f"RMSE: {mean_squared_error(y, y_pred, squared=False):.2f}\n")
+
+
+def round_predictions(predict: np.ndarray, threshold: float = 0.5):
```
Comment (j-adamczyk) [inline training/results.py:22]: Zwracany typ to `np.ndarray`

File: training/results.py [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+from matplotlib import pyplot as plt
+from sklearn.metrics import (
+    ConfusionMatrixDisplay,
+    confusion_matrix,
+    mean_absolute_error,
+    mean_squared_error,
+)
+
+
+def print_check_predictions(y: pd.Series, y_pred: np.ndarray):
```
Comment (j-adamczyk) [inline training/results.py:12]: Zwraca `None`

File: training/results.py [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+    threshold_predict = np.where(
+        (predict % 1) > threshold, np.ceil(predict), np.floor(predict)
+    ).astype("int")
+
+    return np.where(threshold_predict > 20, 21, threshold_predict)
+
+
+def check_round_predictions(
+    round_types: list[float], y: pd.Series, predict: np.ndarray
+):
```
Comment (j-adamczyk) [inline training/results.py:43]: Zwraca `None`

Comment (j-adamczyk) [reply training/results.py:43]: Plus może `evaluate_round_prediction`? Po funkcji z `check` w nazwie spodziewałbym się zwrotu boola

File: training/results.py [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+        print(f"Round type: {threshold}")
+        print_check_predictions(y, threshold_predict)
+
+
+def plot_mae_by_level(
+    y_test: pd.Series,
+    y_pred_test: np.ndarray,
+    title: str = None,
+    figsize: tuple[int, int] = (20, 8),
+):
```
Comment (j-adamczyk) [inline training/results.py:67]: Zwraca `None`

File: training/results.py [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+    plt.show()
+
+
+def plot_confusion_matrix(
+    predict: np.ndarray,
+    y: pd.Series,
+    threshold: float = 0.5,
+    title: str = None,
+    figsize: tuple[int, int] = (10, 10),
+):
```
Comment (j-adamczyk) [inline training/results.py:110]: Zwraca `None`
